### ISSUE TITLE
Filter parser throws wrong exception type for malformed DateTimeOffset literals

### ISSUE THREAD

**Issue by brjohnstmsft** [OData/odata.net#434]
The filter parser should always throw `ODataException` for syntax errors, but sometimes it throws a different exception type. When parsing a filter with a DateTimeOffset literal with no time zone information, it throws `InvalidOperationException`. When parsing a filter with a DateTimeOffset literal with an offset of magnitude greater than 14 hours, it throws `ArgumentOutOfRangeException`. The parser should internally catch these errors and throw the correct `ODataException`.

Here is a full repro:
```csharp
    using System.Collections.Generic;
    using Microsoft.OData.Core;
    using Microsoft.OData.Core.UriParser;
    using Microsoft.OData.Edm;
    using Microsoft.OData.Edm.Library;
    using Microsoft.VisualStudio.TestTools.UnitTesting;

    namespace ODataFilterRepro
    {
        [TestClass]
        public class FilterTests
        {
            [TestMethod]
            [ExpectedException(typeof(ODataException))]
            public void ParseFilterWithMissingTimeZoneExpectODataException()
            {
                CreateParser("createdOn gt 2015-12-29T00:00:00").ParseFilter();
            }

            [TestMethod]
            [ExpectedException(typeof(ODataException))]
            public void ParseFilterWithTimeZoneOutOfRangeExpectODataException()
            {
                CreateParser("createdOn gt 2015-12-29T00:00:00-25:00").ParseFilter();
            }

            private static ODataQueryOptionParser CreateParser(string filter)
            {
                var targetType = new EdmEntityType("test", "document");
                targetType.AddStructuralProperty("createdOn", EdmPrimitiveTypeKind.DateTimeOffset);
                var keyProperty = targetType.AddStructuralProperty("key", EdmPrimitiveTypeKind.String);
                targetType.AddKeys(keyProperty);

                var model = new EdmModel();
                var container = new EdmEntityContainer("test", "default");
                model.AddElement(targetType);
                model.AddElement(container);

                var entitySet = new EdmEntitySet(model.EntityContainer, "docs", targetType);

                var options = new Dictionary<string, string>() { { "$filter", filter } };
                return new ODataQueryOptionParser(model, targetType, entitySet, options);
            }
        }
    }
```

**Comment by chinadragon0515** [comment]
The fix will be,
1. For Platform helper class, it will not throw InvalidOperationException any more, but throw FormatException which is already the exception thrown by XmlConvert.ToDateTimeOffset.
2. In UriParser, when catch FormatException, will check whether the text is similar to datatime offset, and if yes, will throw ODataException without further parsing and more meaningful message.
3. In UriParser, when catch ArgumentOutOfRangeException, will throw ODataException without further parsing and more meaningful message.

**Comment by chinadragon0515** [comment]
Code is merged, close it.

**Comment by brjohnstmsft** [comment]
Thanks! What version of ODataLib will include this fix? 6.15?

**Comment by chinadragon0515** [comment]
Yes, it will be in 6.15.
